Use SHA-256 for hashing PyTorch Dynamo Source class

xingye8 opened the pull request:
Summary:
`Source` class contains a `_hash` field, which is calculated from Python's default `__hash__()` function. The hash value changes across runs because it doesn't use the attributes in the object. 

Since each instance of `Source` object has a different `_hash` field, it causes a downstream issue of generating inconsistent cache key for `AOTAutogradCache`. The cache key partly relies on the content of `AOTConfig`, which saves a list of `Source` in `aot_autograd_arg_pos_to_source`.  Therefore, `AOTAutogradCache` generates a different cache key for the same graph, causing a cache miss.

This diff saves a consistent hash value in the `_hash` field of the `Source` object. The SHA-256 algorithm creates the same hash value if the attributes in the `Source` object are the same.

Test Plan:
```
buck test //mtia/compiler/graph_compiler/test_library:unittest_aot_autograd_cache -- test_dynamic_shapes_different_sizes
```

Differential Revision: D89007953

Skylion007 commented on torch/_guards.py:
```diff
@@ -1142,12 +1144,12 @@ def dataclass_with_cached_hash(
 ) -> type[T] | Callable[[type[T]], type[T]]:
     def wrap(cls_inner: type[T]) -> type[T]:
         new_cls = dataclasses.dataclass(cls_inner, **kwargs)
```
```suggestion
        new_cls = dataclasses.dataclass(cls_inner, hash=False,**kwargs)

xingye8 committed:
Use SHA-256 for hashing PyTorch Dynamo Source class (#170322)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/170322

`Source` class contains a `_hash` field, which is calculated from Python's default `__hash__()` function. The hash value changes across runs because it doesn't use the attributes in the object.

Since each instance of `Source` object has a different `_hash` field, it causes a downstream issue of generating inconsistent cache key for `AOTAutogradCache`. The cache key partly relies on the content of `AOTConfig`, which saves a list of `Source` in `aot_autograd_arg_pos_to_source`.  Therefore, `AOTAutogradCache` generates a different cache key for the same graph, causing a cache miss.

This diff saves a consistent hash value in the `_hash` field of the `Source` object. The SHA-256 algorithm creates the same hash value if the attributes in the `Source` object are the same.

Test Plan:
```
buck test //mtia/compiler/graph_compiler/test_library:unittest_aot_autograd_cache -- test_dynamic_shapes_different_sizes
```

Differential Revision: D89007953
